src/warc2zim/url_rewriting.py

```diff
@@ -152,46 +289,69 @@ def get_without_fragment(url: str) -> str:
+    def __call__(self, item_url: str, *, rewrite_all_url: bool = True) -> str:
         """Rewrite a url contained in a article.
 
         The url is "fully" rewrited to point to a normalized entry path
         """
 
-        url = url.strip()
+        item_url = item_url.strip()
 
-        if url.startswith("data:") or url.startswith("blob:"):
-            return url
+        item_scheme = urlsplit(item_url).scheme
+        if item_scheme and item_scheme not in ("http", "https"):
+            return item_url
 
-        absolute_url = urljoin(self.article_url, url)
+        item_absolute_url = urljoin(self.article_url.value, item_url)
+        item_fragment = urlsplit(item_absolute_url).fragment
 
-        normalized_url = normalize(absolute_url)
+        item_path = normalize(HttpUrl(item_absolute_url))
 
-        if rewrite_all_url or get_without_fragment(normalized_url) in self.known_urls:
-            return self.from_normalized(normalized_url)
+        if rewrite_all_url or item_path in self.existing_zim_paths:
+            return self.get_document_uri(item_path, item_fragment)
         else:
-            logger.debug(f"WARNING {normalized_url} ({url}) not in archive.")
+            if item_path not in self.missing_zim_paths:
+                logger.debug(f"WARNING {item_path} ({item_url}) not in archive.")
+                # maintain a collection of missing Zim Path to not fill the logs with
+                # duplicate messages
+                self.missing_zim_paths.add(item_path)
             # The url doesn't point to a known entry
-            return url
+            return item_absolute_url
 
-    def from_normalized(self, normalized_url_str: str) -> str:
-        normalized_url = urlsplit(f"/{normalized_url_str}")
+    def get_document_uri(self, item_path: ZimPath, item_fragment: str) -> str:
+        """Given an ZIM item path and its fragment, get the URI to use in document
 
-        # relative_to will lost our potential last '/'
-        slash_ending = normalized_url.path[-1] == "/"
-        relative_path = posixpath.relpath(normalized_url.path, self.base_path)
+        This function transforms the path of a ZIM item we want to adress from current
+        document (HTML / JS / ...) and returns the corresponding URI to use.
 
-        if slash_ending:
+        It computes the relative path based on current document location and escape
+        everything which needs to be to transform the ZIM path into a valid RFC 3986 URI
+
+        It also append a potential trailing item fragment at the end of the resulting
+        URI.
+
+        """
+        item_parts = urlsplit(item_path.value)
+
+        # item_path is both path + querystring, both will be url-encoded in the document
+        # so that readers consider them as a whole and properly pass them to libzim
+        item_url = item_parts.path
+        if item_parts.query:
+            item_url += "?" + item_parts.query
+
+        relative_path = posixpath.relpath(
```

Comment:
I don't understand why we're using this os-like API while we use the Path one anywhere else.

```py
PurePosixPath(item_url).relative_to(PurePosixPath(self.article_path.value).parent)
```

Reply:
@mgautierfr could you explain? It is unclear for me as well.

Reply:
I believe @mgautierfr wanted to use POSIX Path (ie. not Windows) because separators in URI paths are aligned with POSIX (I think we discussed this in his original PR) instead of `Path` (that adapts to whatever platform it runs on).
Now whether he used the lower-level `posixpath` on purpose or because he wasn't aware of `PurePosixPath` and the fact that it's our convention is unknown to me

Reply:
I've changed the code to use PurePosixPath now that it can walk up